src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/AbstractDocumentDiagnosticSource.cs

```diff
@@ -2,23 +2,25 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
 using System.Collections.Immutable;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Diagnostics;
-using Microsoft.CodeAnalysis.Shared.Extensions;
-using Microsoft.CodeAnalysis.TaskList;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
 
-internal abstract record AbstractDocumentDiagnosticSource<TDocument>(TDocument Document) : IDiagnosticSource
+internal abstract class AbstractDocumentDiagnosticSource<TDocument> : IDiagnosticSource
```

Comment:
I think these need to implement equality in order for this code to work
https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs,98

Reply:
I see. Seems like it'd make more sense to pass a comparer that compares the documents to RemoveDuplicates.

Reply:
yeah thats fine with me as well

Reply:
Although it would need to compare projects in some cases too (could be duplicate project diagnostic sources in there)

Reply:
Is this in main? I can't find it.

Reply:
Also, all tests are passing, so ... ;)

Reply:
Ah you're right, Cyrus just refactored this.  Looks like we don't need to remove dupes any more since we just go through the possible projects once.
https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs#L172